### PR TITLE
KEP-4008: CRDValidationRatcheting Metrics

### PR DESCRIPTION
**Pull request by alexzielenski**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

Adds metrics to the comparison for CRDValidationRatcheting

#### Which issue(s) this PR fixes:

N/A

#### Special notes for your reviewer:

/sig api-machinery
/assign @apelisse 

This was split off of https://github.com/kubernetes/kubernetes/pull/121343 and is based on top of it, so requires it to merge first.

#### Does this PR introduce a user-facing change?

```release-note
Adds `apiextensions_apiserver_update_ratcheting_time` metric for tracking time taken during requests by feature `CRDValidationRatcheting`
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://kep.k8s.io/4008
```

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/ratcheting_test.go
Comment (logicalhan) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/ratcheting_test.go:336]: An integration test is pretty heavy handed, it's easier to write a unit test for this: https://github.com/kubernetes/kubernetes/blob/6330b2722559c49e76daf347240ba569023257e4/staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics_test.go#L36

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/metrics.go
Comment (logicalhan) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/metrics.go:45]: is 2.5s enough?

Comment (apelisse) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/metrics.go:50]: IMO, 2.5s is basically infinite time yeah. I have very little patience.

Comment (alexzielenski) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/metrics.go:50]: I think so. For the comparison to be on the order of seconds is already very concerning. IMO higher buckets wouldn't give much more information than 2.5s

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve

(from a sig-instrumentation perspective)

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/metrics_test.go
Comment (alexzielenski) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/metrics_test.go:218]: /hold

cc @logicalhan had some difficulties here resetting the histogram metric since it is not a `Vec` and does not expose a `Reset()`. For now the tests are order dependent. 

Do you have any advice to workaround?

**Review by jpbetz — APPROVED** [review]
/approve

The metric def and the bucket distribution LGTM. Leaning on @logicalhan's review of the instrumentation specifics.

**Review by logicalhan — COMMENTED** [review]
/lgtm

for metrics

**Comment by alexzielenski** [comment]
Added another check to the test that the real durations is > 0, and cleaned up naming a little.

/hold cancel

**Comment by dims** [comment]
@logicalhan @jpbetz needs `lgtm` to be re-applied after the last change

**Review by logicalhan — APPROVED** [review]
/lgtm

(for metrics)

**Pull request merged**
